src/internal/obj/bucket.go

```diff
@@ -0,0 +1,122 @@
+type Bucket = blob.Bucket
+
+func amazonHTTPClient() (*http.Client, int, error) {
+	advancedConfig := &AmazonAdvancedConfiguration{}
+	if err := cmdutil.Populate(advancedConfig); err != nil {
+		return nil, -1, errors.Wrap(err, "creating amazon http client")
+	}
+	timeout, err := time.ParseDuration(advancedConfig.Timeout)
+	if err != nil {
+		return nil, -1, errors.Wrap(err, "creating amazon http client")
+	}
+	httpClient := &http.Client{Timeout: timeout}
+	if advancedConfig.NoVerifySSL {
+		transport := http.DefaultTransport.(*http.Transport).Clone()
+		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+		httpClient.Transport = transport
+	}
+	httpClient.Transport = promutil.InstrumentRoundTripper("s3", httpClient.Transport)
+	return httpClient, advancedConfig.Retries, nil
+}
+
+// This seems to be required in order to support disabling ssl verification -- which is needed for EDF testing.
+func amazonSession(ctx context.Context, objURL *ObjectStoreURL) (*session.Session, error) {
+	urlParams, err := url.ParseQuery(objURL.Params)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating amazon session")
+	}
+	endpoint := urlParams.Get("endpoint")
+	// if unset, disableSSL will be false.
+	disableSSL, _ := strconv.ParseBool(urlParams.Get("disableSSL"))
+	region := urlParams.Get("region")
+	httpClient, retries, err := amazonHTTPClient()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating amazon session")
+	}
+	awsConfig := &aws.Config{
+		Region:     aws.String(region),
+		MaxRetries: aws.Int(retries),
+		HTTPClient: httpClient,
+		DisableSSL: aws.Bool(disableSSL),
+		Logger:     log.NewAmazonLogger(ctx),
+	}
+	// Set custom endpoint for a custom deployment.
+	if endpoint != "" {
+		awsConfig.Endpoint = aws.String(endpoint)
+		awsConfig.S3ForcePathStyle = aws.Bool(true)
+	}
+	// Create new session using awsConfig
+	sess, err := session.NewSession(awsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating amazon session")
+	}
+	return sess, nil
+}
+
+// NewAmazonBucket constructs an amazon client by reading credentials
+// from a mounted AmazonSecret. You may pass "" for bucket in which case it
+// will read the bucket from the secret.
+func NewAmazonBucket(ctx context.Context, objURL *ObjectStoreURL) (*Bucket, error) {
+	// Use or retrieve S3 bucket
+	sess, err := amazonSession(ctx, objURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "amazon bucket")
+	}
+	blobBucket, err := s3blob.OpenBucket(ctx, sess, objURL.Bucket, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "amazon bucket")
+	}
+	return blobBucket, nil
+}
+
+// NewBucket creates a Bucket using the given backend and storage root (for
+// local backends).
+// TODO: Not sure if we want to keep the storage root configuration for
+// non-local deployments. If so, we will need to connect it to the object path
+// prefix for chunks.
+func NewBucket(ctx context.Context, storageBackend, storageRoot string) (*Bucket, error) {
```

Comment:
I think we should avoid accessing the OS environment in this function.  It makes it more difficult to reason about the behavior, and it's also not mentioned in the comment above.

Can the storage url be an argument to this function?

Can we move the non-functional behavior into something like `func NewBucketFromEnv(context.Context) (*Bucket, error)` which reads from the environment to get everything?

At that point we would have one easy to test function that works with strings, and then one harder to test function that reads environment variables.  We probably don't need to test a small function that reads environment variables.  The rest of the test suite will exercise it.  We should be able to test the parsing and backend switching logic.

Reply:
We can actually pass in the STORAGE_URL value directly from the pach config. That should eliminate the OS stuff and make this easily testable